tetris_3.py

```diff
@@ -75,6 +54,40 @@ def tetris(d):
     return string
 
 
+def update_board_with_piece(array, entry):
+    piece_array = PIECES[entry['piece']]
+    column = entry['column']
+    al = len(piece_array)  # array length
+    # takes part of a piece individually, not good for Z and S
+    while al >= 1:
+        part = piece_array[al - 1]  # start at the last/bottom part
+        for string in part:
+            pl = len(string)
+        row = 0
+        while row < 21:
+            if all(array[row * 13 + column + space] is ws
+                   for space in range(0, pl)):
+                # if all spaces below the piece are blank, return true
+                row += 1
+            else:
+                array[(row-1)*13 + column] = part
+                for fill in range(1, pl):
+                    array[(row-1)*13 + column + fill] = dl
+                break
+        al -= 1
+
+
+def tetris(pieces_to_drop):
```

Comment:
extracted most of the code out of this function into smaller ones so that it would be easier to analyse. In general you want to have very short, focused functions that do only one thing and do it right (now that we're talking about that, another interesting reading is https://en.wikipedia.org/wiki/Unix_philosophy ('do one thing and do it well')

also:
https://en.wikipedia.org/wiki/Cohesion_(computer_science)
https://en.wikipedia.org/wiki/Coupling_(computer_programming)